src/tools/state/utils/scheduler.ts

```diff
@@ -0,0 +1,48 @@
+import { frame, schedule } from 'timing-functions';
+
+export class Scheduler {
+  // Function that will be run by the scheduler and return an estimate of when
+  // to run it again
+  #runner: () => Promise<number>;
+
+  // Reference to the timeout for the last scheduled run
+  #currentSchedule: number | undefined;
+
+  // Reference to know when the async runner is currently running
+  #currentlyRunning: boolean;
+
+  constructor(runner: () => Promise<number>) {
+    this.#runner = runner;
+    this.#currentlyRunning = false;
+  }
+
+  schedule(delay: number) {
+    console.log(`scheduling in ${delay / 1000}s`);
```

Comment:
```suggestion
```
Just here to help manually test the scheduling logic, will remove before merging